File: Degree_Of_An_Array/degree_of_an_array.py
class Solution:
    def findShortestSubArray(self, nums: List[int]) -> int:
         #[1,2,2,3,1]
        # {1: [0,4], 2: [1,2]}
        # 1: put everything in Hashmap K/V 
        # Key = Actual value
        # Value = list of indices
        # First time you saw it, last time you saw it

        # Degree = whichever Valuelist has the most elems
        # result = value[-1] - value[0]

        

        if not nums:
            return 0

        hashtable = {}
        maxElems = 0
        for i, elem in enumerate(nums):
            
            if not hashtable.get(elem):
                newEntry = [i]
                hashtable[elem] = newEntry

            else:
                hashtable[elem].append(i)

                checkLength = len(hashtable[elem]) 

                if (checkLength > maxElems):
                    maxElems = len(hashtable[elem]) #.values()

        minLength = float('inf')
        for key in hashtable:

            if len(hashtable[key]) == maxElems:

                tempList = hashtable[key]
                currLen = tempList[-1] - tempList[0]

                minLength = currLen if currLen < minLength else minLength

        return 1 if minLength == float('inf') else minLength + 1





    def findShortestSubArray(self, nums: List[int]) -> int:

        # {1: [count, first_index, last_index]}
        # 1: put everything in Hashmap K/V 
        # Key = Actual value
        # Degree = count
        # values = [count, first_index, last_index]

        if not nums:
            return 0

        hashtable = {}
        maxElems = 0
        for i, elem in enumerate(nums):
            
            if not hashtable.get(elem):
                newEntry = [1, i]
                hashtable[elem] = newEntry

            else:
                hashtable[elem][0] += 1
                if len(hashtable[elem]) == 2:
                    hashtable[elem].append(i)
                hashtable[elem][2] = i
                maxElems = hashtable[elem][0] if hashtable[elem][0] > maxElems else maxElems

        minLength = float('inf')
        if maxElems > 1:
            for key in hashtable:

                if hashtable[key][0] == maxElems:

                    tempList = hashtable[key]
                    currLen = tempList[2] - tempList[1]

                    minLength = currLen if currLen < minLength else minLength

        return 1 if minLength == float('inf') else minLength + 1

File: Degree_Of_An_Array/degree_of_an_array.cpp
#include <unordered_map>
#include <vector>
#include <limits>
#include <stdio.h>

class Solution {
public:
    int findShortestSubArray(vector<int>& nums) {
        if (nums.empty()) {
            return 0;
        }
        std::unordered_map<int, std::vector<int>> hashMap;
        int maxElems = 0;

        for (int i = 0; i < nums.size(); i++) {
            int elem = nums[i];

            if (hashMap.count(elem) == 0) {

                std::vector<int> newEntry = {1, i, i};
                hashMap[elem] = newEntry;
            } else {
 
  
                hashMap[elem][0]++;
                hashMap[elem][2] = i;
            }
            maxElems = hashMap[elem][0] > maxElems ? hashMap[elem][0] : maxElems;
            
        }
        int minLength = std::numeric_limits<int>::max();

        if (maxElems == 1) {
            return maxElems;
        }
        for (auto it = hashMap.begin(); it != hashMap.end(); ++it ) {
            const std::vector<int>& values = it->second;

            if (values[0] == maxElems) {
                int currLen = values[2] - values[1] + 1;
                minLength = currLen < minLength ? currLen : minLength;
            }

        }
        return minLength;
    }
};